### PR TITLE
treewide: use seastar::format() or fmt::format() explicitly

### PR DESCRIPTION
**Pull request by tchaikov**
before this change, we rely on `using namespace seastar` to use `seastar::format()` without qualifying the `format()` with its namespace. this works fine until we changed the parameter type of format string `seastar::format()` from `const char*` to `fmt::format_string<...>`. this change practically invited `seastar::format()` to the club of `std::format()` and `fmt::format()`, where all members accept a templated parameter as its `fmt` parameter. and `seastar::format()` is not the best candidate anymore. as, because argument-dependent lookup (ADT for short) favors the function which is in the same namespace as its parameter, but what if one of the argument is in `std` namespace, while another in `seastar`? that is what is happening scylladb in quite a few caller sites of `format()`, hence ADT is not able to tell which function in the winner in the name lookup competition:

```
/__w/scylladb/scylladb/mutation/mutation_fragment_stream_validator.cc:265:12: error: call to 'format' is ambiguous
  265 |     return format("{} ({}.{} {})", _name_view, s.ks_name(), s.cf_name(), s.id());
      |            ^~~~~~
/usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../include/c++/14/format:4290:5: note: candidate function [with _Args = <const std::basic_string_view<char> &, const seastar::basic_sstring<char, unsigned int, 15> &, const seastar::basic_sstring<char, unsigned int, 15> &, const utils::tagged_uuid<table_id_tag> &>]
 4290 |     format(format_string<_Args...> __fmt, _Args&&... __args)
      |     ^
/__w/scylladb/scylladb/seastar/include/seastar/core/print.hh:143:1: note: candidate function [with A = <const std::basic_string_view<char> &, const seastar::basic_sstring<char, unsigned int, 15> &, const seastar::basic_sstring<char, unsigned int, 15> &, const utils::tagged_uuid<table_id_tag> &>]
  143 | format(fmt::format_string<A...> fmt, A&&... a) {
      | ^
```

in this change, we

change all `format()` to either `fmt::format()` or `seastar::format()` with following rules:
- if the caller expects an `sstring` or `std::string_view`, change to  `seastar::format()`
- if the caller expects an `std::string`, change to `fmt::format()`.  because, `sstring::operator std::basic_string` would incur a deep copy.

we will need another change to enable scylladb to compile with the latest seastar. namely, to pass the format string as a templated parameter down to helper functions which format their parameters. to miminize the scope of this change, let's include that change when bumping up the seastar submodule. as that change will depend on the seastar change.

---

this change addresses a build failure due a new change in seastar, hence no impacts on the production. so no need to backport.

### CODE REVIEW

**Comment by tchaikov** [comment]
this change should address the build failure identified by https://github.com/scylladb/scylladb/actions/runs/10695413445/job/29648811715

**Review by denesb — COMMENTED** [review]
What will happen when someone, inevitably, introduces a new unqualified call to `format()`? The build will fail?

**Comment by tchaikov** [comment]
> What will happen when someone, inevitably, introduces a new unqualified call to `format()`? The build will fail?

it depends, please consider following program:
```c++
#include <iostream>
#include <string>
#include <fmt/format.h>
#include <seastar/core/print.hh>

using namespace seastar;

int main() {
    
    std::cout << format("{}", 1);
    std::cout << format("{}", std::string{}); # ambiguous
    std::cout << format("{} {}", string{}, sstring{}); # ambiguous
}
```

- the first `format()` does not fail the build. as `seastar::format()` is the only winner. `1` is not in the namespace of `std.
- the second `format()` fails the build. as ADT can find 

  - `std::format()` -- `string` is in `std`.and 
  - `seastar::format()` -- it matches anyway
- the third one fails the build. for similar reason.

**Comment by nyh** [comment]
> What will happen when someone, inevitably, introduces a new unqualified call to `format()`? The build will fail?

Could we introduce somewhere ::format() or some other trick that will make all namespace-less calls to `format()`  fail compilation in a more-or-less easy to understand way?

File: auth/authentication_options.hh
Comment (nyh) [inline auth/authentication_options.hh:30]: Why was it necessary to move this code in this patch?

**Review by nyh — CHANGES_REQUESTED** [review]
I stopped reviewing this patch in the middle, because I'm worried I didn't understand something:

In 99% of the cases (maybe even 100%?), we want format() to return a std::string, not seastar::sstring. I think this means you want all of them to use fmt::format, **not** seastar::format. Yet you seemed to use seastar::format in a lot of places I don't understand. I don't know if those were all mistakes, or there is some other reason.

After seeing dozens of (what I think are) mistakes in your patch using seastar::format instead of fmt::format, I'm starting to wonder: Could we just do "using format = fmt::format", and in the very rare cases (if there are even any such cases... I couldn't find even one) where we need seastar::format we could use that? This will make this patch thousands of lines shorter, and the code shorter (no "fmt::" everywhere).

File: alternator/auth.cc
Comment (tchaikov) [reply alternator/auth.cc:37]: `format` is not a type, it is a function. so we cannot use 
```c++
using format = fmt::format;
```
to define an alias for it. what we can though is to use
```c++
using fmt::format;
```
to expose it in the current namespace. but this does not work, because we already have
```c++
using namespace seastar;
```
which puts all the names in `seastar` in the global namespace. so we already have the equivalent declaration in scylla.

> But on the other hand, maybe it's better to always use namespaces - like we do for "std" - so your longer patch has merit.

i don't intend / dare to change the fact of `using namespace seastar`. it's unusual in an application. but seems we do this on purpose. and it's not in the scope of this change. i just wanted fix the build with the latest seastar, ASAP. to reduce the surprise when some of us want to bump the seastar submodule in (near) future. that'd be frustrating facing ton of errors, when he/she just want to bring a fix in seastar to scylla.

**Comment by tchaikov** [comment]
Nadav, thank you very much for the review and the insights. i appreciate them.

> I stopped reviewing this patch in the middle, because I'm worried I didn't understand something:
> 
> In 99% of the cases (maybe even 100%?), we want format() to return a std::string, not seastar::sstring.

could be 96.5%. 

> I think this means you want all of them to use fmt::format, **not** seastar::format. Yet you seemed to use seastar::format in a lot of places I don't understand. I don't know if those were all mistakes, or there is some other reason.

they are leftovers.

i have a handful commits in my local branch. all of them intend to address the build failure with the `seastar::format()` change. they've been hanging around for months. in the very beginning, i was just changing `format()` to `seastar::format()`, as it is exactly what we had. but when time passed by, i realized that, under certain circumstances, i should do better than preserving the existing behavior by saving the cost of deep copy, so i changed the policy and updated a couple places to use `fmt::format()` instead of `seastar::format()` , but was preempted by some other tasks. so as you found out, i failed to change all of them.

> 
> After seeing dozens of (what I think are) mistakes in your patch using seastar::format instead of fmt::format, I'm starting to wonder: Could we just do "using format = fmt::format", and in the very rare cases (if there are even any such cases... I couldn't find even one) where we need seastar::format we could use that? This will make this patch thousands of lines shorter, and the code shorter (no "fmt::" everywhere).

i agree with you on the thought. but as explained above. I cannot find a way to help the compiler to disambiguate  these two functions. even if we use 
```c++
using fmt::format;
```
or 
```c++
using std::format;
```
this just gets them into the global naming space.  however, argument-dependent lookup still prioritizes `std::format()`, if its argument(s) is in `std`. considering this, `std::format()` and `seastar::format()` are in competition. as a result, the compiler cannot determine which one is preferable.

what i can think of, is to rename `seastar::format()` to something else or just deprecate and then drop it by guarding it with a macro.  and let ADL to pick `std::format()` for us, like how `std::operator<<` is picked. see https://godbolt.org/z/85b9aKPof . but that'd be a breaking change in seastar, we might need to consider it in a different venue.

in the meanwhile, i am auditing all the changed places, and fixing them case-by-case. but before #20447 lands, i am marking this change as a draft.

**Comment by nyh** [comment]
Your answer left me even more confused.... Given there is both fmt::format and std::format, why or when should we use the former (fmt) instead of the latter (std)? I realize that fmt has some fancy extensions not found in std, but do we really need those in 99% of the calls which just format silly error messages?

I was never happy that we use `using namespace seastar;' in Scylla source code, and didn't recommend this practice in the Seastar tutorial, but a lot of good that did me ;-) I was hoping there was a way to "undo" this "using namespace seastar" for a specific function, or that "using format = ..." would override the less specific "using namespace seastar", but I guess I was overly optimistic (I didn't actually try it).

**Comment by tchaikov** [comment]
> Your answer left me even more confused.... Given there is both fmt::format and std::format, why or when should we use the former (fmt) instead of the latter (std)? 

this is indeed a good question. but i don't have a good answer -- when we are confident that `std::format()` is able to fulfill our needs. 

> I realize that fmt has some fancy extensions not found in std, but do we really need those in 99% of the calls which just format silly error messages?

i'd do further research before answering your question. but i agree, since `std::format()` is available in libstdc++, we'll consider switching to it under certain circumstances in future. however, given the scope of this change, we'll focus on the immediate task -- enable us to compile with the latest seastar submodule sooner than later.  i preferred to implement changes in smaller steps.
 
> 
> I was never happy that we use `using namespace seastar;' in Scylla source code, and didn't recommend this practice in the Seastar tutorial, but a lot of good that did me ;-) I was hoping there was a way to "undo" this "using namespace seastar" for a specific function, or that "using format = ..." would override the less specific "using namespace seastar", but I guess I was overly optimistic (I didn't actually try it).

yeah, anyway, switching to `seastar::format()` was not an attempt to address this (problem), if we consider it a problem.

**Comment by tchaikov** [comment]
v2:

- audited all the changed places, and fixing them case-by-case so that we use `fmt::format()` at all places where `std::string` is expected

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11421/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11421/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3464/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11487/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/414/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11421/)
> #### Failed Tests (1/34984):
> 
>     * [test_tablet_drain_failure_during_decommission.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11421/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_dev_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.4)
> 
> 
> #### Build Details:
> 
>     * Duration: 3 hr 9 min
> 
>     * Builder: spider7.cloudius-systems.com

#20060

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11430/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11430/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/419/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3469/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11492/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11430/)
> #### Failed Tests (2/35019):
> 
>     * [test_parallel_operations_for_10ks_and_10tables_and_clearallsnapshots](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11430/testReport/junit/snapshot_test/TestParallelSnapshotOperations/Tests___dtest_with_tablets___test_parallel_operations_for_10ks_and_10tables_and_clearallsnapshots) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_parallel_operations_for_10ks_and_10tables_and_clearallsnapshots)

#20268

> 
>     * [test_replace.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11430/testReport/junit/%28root%29/test_tablets_removenode/Tests___Unit_Tests___test_replace_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace.debug.1)

#20492

> 
> 
> #### Build Details:
> 
>     * Duration: 3 hr 9 min
> 
>     * Builder: spider2.cloudius-systems.com

**Comment by tchaikov** [comment]
v3:

- rebased against master HEAD.

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11477/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11477/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/456/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11528/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3508/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11477/)
> #### Failed Tests (1/35096):
> 
>     * [test_tablet_drain_failure_during_decommission.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11477/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_dev_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.4)
> 
> 
> #### Build Details:
> 
>     * Duration: 3 hr 15 min
> 
>     * Builder: spider2.cloudius-systems.com

#20060

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11484/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11484/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/462/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11534/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3514/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11484/)
> #### Build Details:
> 
>     * Duration: 6 hr 18 min
> 
>     * Builder: i-07d032bb753a91624 (m5ad.8xlarge)

> [FullDtest / full-split001 / snapshot_test.TestParallelSnapshotOperations.test_parallel_operations_for_10ks_and_10tables_and_clearallsnapshots](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/462/testReport/junit/snapshot_test/TestParallelSnapshotOperations/FullDtest___full_split001___test_parallel_operations_for_10ks_and_10tables_and_clearallsnapshots/)

#20268

>     [FullDtest / full-split004 / data_distribution_balance_test.TestDataDistribution.test_data_distribution_balance[4-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/462/testReport/junit/data_distribution_balance_test/TestDataDistribution/FullDtest___full_split004___test_data_distribution_balance_4_TimeWindowCompactionStrategy_/)

#20503

>     [FullDtest / full-split004 / nodetool_additional_test.TestNodetool.test_concurrent_repair](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/462/testReport/junit/nodetool_additional_test/TestNodetool/FullDtest___full_split004___test_concurrent_repair/)

#20533

**Comment by tchaikov** [comment]
@scylladb/scylla-maint and @nyh  hello maintainers and Nadav, could you help review this change?

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ❌ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11511/)
> #### Build Failure:
> 
> ```
> [2024-09-11T05:30:14.178Z] FAILED: build/debug/readers/multishard.o 
> [2024-09-11T05:30:14.178Z] clang++ -MD -MT build/debug/readers/multishard.o -MF build/debug/readers/multishard.o.d -std=c++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/gen/include -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEBUG_PROMISE -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -I/usr/include/p11-kit-1 -DWITH_GZFILEOP -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_BUILD_MODE=debug -g -gz -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/debug/gen -std=gnu++23  -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-psabi -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN  -c -o build/debug/readers/multishard.o readers/multishard.cc
> [2024-09-11T05:30:14.182Z] readers/multishard.cc:202:50: error: unexpected namespace name fmt: expected expression
> [2024-09-11T05:30:14.183Z]   202 |         on_internal_error(mrlog, seastar::format(fmt, arg...));
> [2024-09-11T05:30:14.183Z]       |                                                  ^
> [2024-09-11T05:30:14.184Z] 1 error generated.
> ```
> 
> #### Build Details:
> 
>     * Duration: 40 min
> 
>     * Builder: spider2.cloudius-systems.com

the build failure was due to a bad rebase. fixed in the latest revision.

File: alternator/auth.cc
Comment (avikivity) [reply alternator/auth.cc:37]: > i don't intend / dare to change the fact of using namespace seastar. it's unusual in an application. but seems we do this on purpose. 

Seastar/ScyllaDB were born as one project (then named Urchin). Later Seastar was extracted from it. b958a443044c7182cc5a92ea49d78bdaa337ea7e is shared with the seastar repo. Only later did we add the seastar namespace.

Comment (avikivity) [reply alternator/auth.cc:37]: btw, we could use fmt::format() if we use std::string instead of sstring. There's an option to alias sstring to std::string, but IIRC it brings a performance regression. If we drop our dependency on strings in the fast path (lookup keyspace, what else?) we could standardize on std::string.

**Comment by tchaikov** [comment]
included by 3e84d43f9348bde011220353539c11a6ed29e291. thanks Avi.

closing.

**Pull request closed without merging**
